### ISSUE TITLE
Cannot invoke "--sudo" with cargo pgrx install - "sudo_install" not found

### ISSUE THREAD

**Issue by bavalpey** [pgcentralfoundation/pgrx#1479]
Trying to follow the documentation for installing an extension with ``cargo pgrx install --sudo``
Every time I do so, I receive the following error messages:
```
Error:
   0: No such file or directory (os error 2)

Location:
   /home/bavalpey/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cargo-pgrx-0.11.2/src/command/sudo_install.rs:72
```

Looks like the path for this doesn't exist in the installation.  Is there another path this should be in, or did I not install something?
I installed pgrx via ``cargo install --locked cargo-pgrx`` with no other flags.

**Comment by eeeebbbbrrrr** [comment]
~~is `sudo` on your $PATH?  It needs to be in order for this to work.~~

**Comment by bavalpey** [comment]
> is `sudo` on your $PATH? It needs to be in order for this to work.

yes

**Comment by eeeebbbbrrrr** [comment]
Can you please show us the entire output of `cargo pgrx install --sudo`?

**Comment by bavalpey** [comment]
```
RUST_BACKTRACE=1 cargo pgrx install --sudo --release
       Using PgConfig("pg15") and `pg_config` from /usr/bin/pg_config
    Building extension with features pg15
     Running command "/home/bavalpey/.rustup/toolchains/1.75-x86_64-unknown-linux-gnu/bin/cargo" "build" "--release" "--features" "pg15" "--no-default-features" "--message-format=json-render-diagnostics"
    Finished release [optimized] target(s) in 0.07s
  Installing extension
     Copying control file to target/release/pg_jsonschema-pg15/usr/share/postgresql/15/extension/pg_jsonschema.control
     Copying shared library to target/release/pg_jsonschema-pg15/usr/lib/postgresql/15/lib/pg_jsonschema.so
 Discovering SQL entities
  Discovered 4 SQL entities: 0 schemas (0 unique), 4 functions, 0 types, 0 enums, 0 sqls, 0 ords, 0 hashes, 0 aggregates, 0 triggers
     Writing SQL entities to target/release/pg_jsonschema-pg15/usr/share/postgresql/15/extension/pg_jsonschema--0.2.0.sql
    Finished installing pg_jsonschema

Using sudo to copy extension files from /home/bavalpey/repos/pg_jsonschema/target/release/pg_jsonschema-pg15
Error:
   0: No such file or directory (os error 2)

Location:
   /home/bavalpey/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cargo-pgrx-0.11.2/src/command/sudo_install.rs:72

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

   0: cargo_pgrx::command::install::execute
      at /home/bavalpey/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cargo-pgrx-0.11.2/src/command/install.rs:63

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 3 frames hidden ⋮
   4: <cargo_pgrx::command::sudo_install::SudoInstall as cargo_pgrx::CommandExecute>::execute::h484f2c54276e7703
      at <unknown source file>:<unknown line>
   5: <cargo_pgrx::command::install::Install as cargo_pgrx::CommandExecute>::execute::hf3f0394d384787a7
      at <unknown source file>:<unknown line>
   6: <cargo_pgrx::command::pgrx::CargoPgrxSubCommands as cargo_pgrx::CommandExecute>::execute::h67e0932fe72dfde3
      at <unknown source file>:<unknown line>
   7: cargo_pgrx::main::hc0343a0431421338
      at <unknown source file>:<unknown line>
   8: std::sys_common::backtrace::__rust_begin_short_backtrace::h892b586d37d82bb2
      at <unknown source file>:<unknown line>
   9: std::rt::lang_start::{{closure}}::ha6e7c98fa9c7c965
      at <unknown source file>:<unknown line>
  10: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hd95060ecd5e1ca24
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:284
  11: std::panicking::try::do_call::h6e8cf51db32a6e4b
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552
  12: std::panicking::try::h3a52eefe24fe3c29
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516
  13: std::panic::catch_unwind::h24c28c23c02c3841
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142
  14: std::rt::lang_start_internal::{{closure}}::h705d3c9cbc06ef47
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148
  15: std::panicking::try::do_call::ha21f52ba13158470
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552
  16: std::panicking::try::h5581346bf6aeb1f8
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516
  17: std::panic::catch_unwind::h7919645a6b72e25b
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142
  18: std::rt::lang_start_internal::h12de51168669836e
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148
  19: std::rt::lang_start::ha2f352f872bf3d86
      at <unknown source file>:<unknown line>
  20: __libc_start_call_main<unknown>
      at ./csu/../sysdeps/nptl/libc_start_call_main.h:58
  21: __libc_start_main_impl<unknown>
      at ./csu/../csu/libc-start.c:392
  22: _start<unknown>
      at <unknown source file>:<unknown line>
```

**Comment by eeeebbbbrrrr** [comment]
does this path exist?  `/home/bavalpey/repos/pg_jsonschema/target/release/pg_jsonschema-pg15`?

**Comment by eeeebbbbrrrr** [comment]
And what about:  `/usr/share/postgresql/15/extension/`

**Comment by bavalpey** [comment]
Yes, both of those paths exist.

**Comment by bavalpey** [comment]
``/home/bavalpey/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cargo-pgrx-0.11.2/src/command/`` ~~does not~~ does as well

**Comment by eeeebbbbrrrr** [comment]
And what about `/usr/lib/postgresql/15/lib/`?

**Comment by bavalpey** [comment]
Yes, ``/usr/lib/postgresql/15/lib`` (and ``/bin``) exists

**Comment by eeeebbbbrrrr** [comment]
ahh.  We have a fix for this in the `develop` branch, but it hasn't been released yet.

**Comment by eeeebbbbrrrr** [comment]
It came in as 04351c2b4ba66f1c7511356efbb41a8272dff29e on 10/24/23 but never founds its way into the v0.11.x series.  :(

Lets me and @workingjubilee discuss and figure out if we want to do a v0.11.3 or not.  I think Jubilee was thinking about a v0.12-alpha, but you probably don't want that.

**Comment by bavalpey** [comment]
Ah.  Do you have a workaround right now? Honestly I just want to install pg_jsonschema for my local postgres version.  Maybe I can copy the files directly with sudo, and then somehow register to pgrx that the extension has been installed?

**Comment by eeeebbbbrrrr** [comment]
I think the bug was that the code wants the actual destination **files** to already exist.  So you might could `sudo touch ...` those files once and then you'd be good.  dumb bug

**Comment by bavalpey** [comment]
Wait, if I just patch up my local copy of ``sudo_install.rs`` with that fix, that should work, no?

**Comment by bavalpey** [comment]
> I think the bug was that the code wants the actual destination **files** to already exist. So you might could `sudo touch ...` those files once and then you'd be good. dumb bug

Do you know which specific files?

**Comment by eeeebbbbrrrr** [comment]
> Wait, if I just patch up my local copy of `sudo_install.rs` with that fix, that should work, no?

Maybe?

**Comment by eeeebbbbrrrr** [comment]
> > I think the bug was that the code wants the actual destination **files** to already exist. So you might could `sudo touch ...` those files once and then you'd be good. dumb bug
> 
> Do you know which specific files?

These files, but in their final destination in `/usr/...`

```
     Copying control file to target/release/pg_jsonschema-pg15/usr/share/postgresql/15/extension/pg_jsonschema.control
     Copying shared library to target/release/pg_jsonschema-pg15/usr/lib/postgresql/15/lib/pg_jsonschema.so
```

**Comment by bavalpey** [comment]
Curious, are these command files compiled on the fly? Why do they exist as ``.rs``?

**Comment by eeeebbbbrrrr** [comment]
ie, maybe you can do:

```shell
$ sudo touch /usr/share/postgresql/15/extension/pg_jsonschema.control /usr/lib/postgresql/15/lib/pg_jsonschema.so
```

once and then `cargo pgrx install --sudo` will work.

**Comment by eeeebbbbrrrr** [comment]
> Curious, are these command files compiled on the fly? Why do they exist as `.rs`?

I think you have some misunderstanding of the files in the `~/.cargo/registry` directory.

You're running the `~/.cargo/bin/cargo-pgrx` binary, which is already compiled.

Try the `sudo touch` thing.  I really don't want to recommend you start locally hacking our code.

**Comment by bavalpey** [comment]
> Try the `sudo touch` thing. I really don't want to recommend you start locally hacking our code.
```
RUST_BACKTRACE=1 cargo pgrx install --sudo --release
       Using PgConfig("pg15") and `pg_config` from /usr/bin/pg_config
    Building extension with features pg15
     Running command "/home/bavalpey/.rustup/toolchains/1.75-x86_64-unknown-linux-gnu/bin/cargo" "build" "--release" "--features" "pg15" "--no-default-features" "--message-format=json-render-diagnostics"
    Finished release [optimized] target(s) in 0.05s
  Installing extension
     Copying control file to target/release/pg_jsonschema-pg15/usr/share/postgresql/15/extension/pg_jsonschema.control
     Copying shared library to target/release/pg_jsonschema-pg15/usr/lib/postgresql/15/lib/pg_jsonschema.so
 Discovering SQL entities
  Discovered 4 SQL entities: 0 schemas (0 unique), 4 functions, 0 types, 0 enums, 0 sqls, 0 ords, 0 hashes, 0 aggregates, 0 triggers
     Writing SQL entities to target/release/pg_jsonschema-pg15/usr/share/postgresql/15/extension/pg_jsonschema--0.2.0.sql
    Finished installing pg_jsonschema

Using sudo to copy extension files from /home/bavalpey/repos/pg_jsonschema/target/release/pg_jsonschema-pg15
       Running sudo cp /home/bavalpey/repos/pg_jsonschema/target/release/pg_jsonschema-pg15/usr/share/postgresql/15/extension/pg_jsonschema.control /usr/share/postgresql/15/extension/pg_jsonschema.control
       Running sudo cp /home/bavalpey/repos/pg_jsonschema/target/release/pg_jsonschema-pg15/usr/lib/postgresql/15/lib/pg_jsonschema.so /usr/lib/postgresql/15/lib/pg_jsonschema.so
Error:
   0: No such file or directory (os error 2)

Location:
   /home/bavalpey/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cargo-pgrx-0.11.2/src/command/sudo_install.rs:72

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

   0: cargo_pgrx::command::install::execute
      at /home/bavalpey/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cargo-pgrx-0.11.2/src/command/install.rs:63

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                ⋮ 3 frames hidden ⋮
   4: <cargo_pgrx::command::sudo_install::SudoInstall as cargo_pgrx::CommandExecute>::execute::h484f2c54276e7703
      at <unknown source file>:<unknown line>
   5: <cargo_pgrx::command::install::Install as cargo_pgrx::CommandExecute>::execute::hf3f0394d384787a7
      at <unknown source file>:<unknown line>
   6: <cargo_pgrx::command::pgrx::CargoPgrxSubCommands as cargo_pgrx::CommandExecute>::execute::h67e0932fe72dfde3
      at <unknown source file>:<unknown line>
   7: cargo_pgrx::main::hc0343a0431421338
      at <unknown source file>:<unknown line>
   8: std::sys_common::backtrace::__rust_begin_short_backtrace::h892b586d37d82bb2
      at <unknown source file>:<unknown line>
   9: std::rt::lang_start::{{closure}}::ha6e7c98fa9c7c965
      at <unknown source file>:<unknown line>
  10: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hd95060ecd5e1ca24
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:284
  11: std::panicking::try::do_call::h6e8cf51db32a6e4b
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552
  12: std::panicking::try::h3a52eefe24fe3c29
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516
  13: std::panic::catch_unwind::h24c28c23c02c3841
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142
  14: std::rt::lang_start_internal::{{closure}}::h705d3c9cbc06ef47
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148
  15: std::panicking::try::do_call::ha21f52ba13158470
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552
  16: std::panicking::try::h5581346bf6aeb1f8
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516
  17: std::panic::catch_unwind::h7919645a6b72e25b
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142
  18: std::rt::lang_start_internal::h12de51168669836e
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148
  19: std::rt::lang_start::ha2f352f872bf3d86
      at <unknown source file>:<unknown line>
  20: __libc_start_call_main<unknown>
      at ./csu/../sysdeps/nptl/libc_start_call_main.h:58
  21: __libc_start_main_impl<unknown>
      at ./csu/../csu/libc-start.c:392
  22: _start<unknown>
      at <unknown source file>:<unknown line>
```

**Comment by bavalpey** [comment]
So, a different error, now.

**Comment by eeeebbbbrrrr** [comment]
sorry, one more file:  `/usr/share/postgresql/15/extension/pg_jsonschema--0.2.0.sql`

**Comment by bavalpey** [comment]
Ok, this worked, thank you so much for holding my hand to get this to work ._.

**Comment by eeeebbbbrrrr** [comment]
🤝

**Comment by eeeebbbbrrrr** [comment]
sorry 'bout the bug.  these things happen.  pgrx is... complex.

**Comment by bavalpey** [comment]
> sorry 'bout the bug. these things happen. pgrx is... complex.

Not at all, I completely understand! You have plenty of disclaimers that pgrx is complex, and it doesn't bother me in the slightest.

**Comment by workingjubilee** [comment]
This should be addressed in 0.11.3
